### PR TITLE
feat(CatalogTable): truncate long description with ellipsis and hoverover tooltip

### PR DESCRIPTION
**Pull request by kiranpatel11**
## Hey, I just made a Pull Request!

Fixes backstage/backstage#4297
 
> ⚠️ KNOWN BUG : When switching between Owned and All Services, tooltip order gets messed up. Something to do with registration of EventListener in OverflowTooltip component. **Need help from react experts to fix it.**

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added storybook component
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
CatalogTable

### CODE REVIEW

**Comment by kiranpatel11** [comment]
Hi @adamdmharvey,

I have added reusable OverflowTooltip component for supporting truncation of longer tax with conditional tooltip for longer text. And used the component in CatalogTable and ApiExplorerTable.

There are few minor issues, where I will need help from react experts:
1. When switching between Owned and All, the dom elements get rerendered and it messes up the order and behavior of tooltip.
2. If the text is truncated by just 1 chars(or may be 2), the tooltip does not seem to appear.

File: packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  // compare once and add resize listener on "componentDidMount"
+  useEffect(() => {
+    compareSize();
+    window.addEventListener('resize', compareSize);
+  }, []);
+
+  // remove resize listener again on "componentWillUnmount"
+  useEffect(
+    () => () => {
+      window.removeEventListener('resize', compareSize);
```
Comment (freben) [inline packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx:49]: You just return this function from the useEffect above, no need to make two of them. Also, move the compareSize fn inside the useEffect too in my opinion, keeps everything contained in one place and doesn't go through the trouble of recreating that function unnecessarily on every draw. Adding setHover as a dependency of the effect is fine if the linter complains, but I think it won't

**Review by freben — COMMENTED** [review]
It's a pity to have to specify a certain width, since it'll be hard to get right on all screen sizes etc. But it's the best we can do if we want to use ellipsis right? Due to css limitations

File: packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx (hunk 1 above)
Comment (kiranpatel11) [reply packages/core/src/components/OverflowTooltip/OverflowTooltip.tsx:49]: Agreed specific width is far from ideal, at least it should be calculated dynamically. Need help from css expert.

on the compareSize will try with single useEffect.

**Comment by kiranpatel11** [comment]
> It's a pity to have to specify a certain width, since it'll be hard to get right on all screen sizes etc. But it's the best we can do if we want to use ellipsis right? Due to css limitations

it does work with max-width, is that an option ? I am not a css expert, and need help here.

**Comment by adamdmharvey** [comment]
A few useless Stack Overflow findings out there, but this looked interesting:

https://github.com/amirfefer/react-ellipsis-with-tooltip

Done little to no research, so not suggesting a replacement, but could be investigated to see if a variety of techniques might work.  (e.g., include this package into the Backstage core, or use concepts as defined in that library in as a extension to the Backstage core by enhancing OOTB Material-UI `ToolTip`, who knows!!!)

**Comment by freben** [comment]
> A few useless Stack Overflow findings out there, but this looked interesting:

That one uses Bootstrap rather than MUI. But maybe the code of the component can be inspiring!

**Comment by kiranpatel11** [comment]
> A few useless Stack Overflow findings out there, but this looked interesting:
> 
> https://amirfefer.github.io/react-ellipsis-with-tooltip
> 
> https://github.com/amirfefer/react-ellipsis-with-tooltip
> 
> Done little to no research, so not suggesting a replacement, but could be investigated to see if a variety of techniques might work. (e.g., include this package into the Backstage core, or use concepts as defined in that library in as a extension to the Backstage core by enhancing OOTB Material-UI `ToolTip`, reverse engineer the CSS/methods they used for your PR, who knows!!!)

the storybook example of this one, does have fixed width on the container. Issue in our case is, the width of the TableCell/TableColumn is calculated dynamically based on the content.

**Comment by kiranpatel11** [comment]
closing due to merge gone wrong...

**Pull request closed without merging**